src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1248 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
+    /**
+     * Looks for a {@link ItemPickaxe} to mine a block of {@code requiredLevel},
+     * in the given {@link ICapabilityProvider}. The {@link ItemPickaxe} tool
+     * level cannot exceed the given {@code maximalLevel}.
+     *
+     * @param provider      {@link ICapabilityProvider} to check in.
+     * @param requiredLevel The minimal required {@link ItemPickaxe} level
+     * @param maximalLevel  The maximal tool level of the {@link ItemPickaxe}
+     * @return True if the {@link ICapabilityProvider} contains a {@link
+     * ItemPickaxe} with the given minimal required level.
+     */
+    public static boolean isPickaxeInProvider(ICapabilityProvider provider, final int requiredLevel, final int maximalLevel)
+    {
+        return hasItemInProvider(provider, (ItemStack stack) -> Utils.checkIfPickaxeQualifies(requiredLevel, Utils.getMiningLevel(stack, Utils.PICKAXE))
+                                                                  && InventoryUtils.hasToolLevel(stack, Utils.PICKAXE, maximalLevel));
+    }
+
+    /**
+     * Adapted from {@link net.minecraft.entity.player.InventoryPlayer#addItemStackToInventory(ItemStack)}.
+     *
+     * @param provider  {@link ICapabilityProvider} to add itemstack to.
+     * @param itemStack ItemStack to add.
+     * @return True if successful, otherwise false.
+     */
+    public static boolean addItemStackToProvider(@NotNull final ICapabilityProvider provider, @Nullable ItemStack itemStack)
+    {
+        return getItemHandlersFromProvider(provider).stream().anyMatch(handler -> addItemStackToItemHandler(handler, itemStack));
+    }
+
+    /**
+     * Adapted from {@link net.minecraft.entity.player.InventoryPlayer#addItemStackToInventory(ItemStack)}.
+     *
+     * @param provider  {@link ICapabilityProvider} to add itemstack to.
+     * @param itemStack ItemStack to add.
+     * @return True if successful, otherwise false.
+     */
+    public static ItemStack addItemStackToProviderWithResult(@NotNull final ICapabilityProvider provider, @Nullable ItemStack itemStack)
+    {
+        ItemStack activeStack = itemStack;
+
+        if (isItemStackEmpty(activeStack))
+        {
+            return EMPTY;
+        }
+
+        for (IItemHandler handler : getItemHandlersFromProvider(provider))
+        {
+            activeStack = addItemStackToItemHandlerWithResult(handler, activeStack);
+        }
+
+        return activeStack;
+    }
+
+    /**
+     * Adapted from {@link net.minecraft.entity.player.InventoryPlayer#addItemStackToInventory(ItemStack)}.
+     *
+     * @param itemHandler {@link IItemHandler} to add itemstack to.
+     * @param itemStack   ItemStack to add.
+     * @return True if successful, otherwise false.
+     */
+    public static ItemStack addItemStackToItemHandlerWithResult(@NotNull final IItemHandler itemHandler, @Nullable ItemStack itemStack)
+    {
+        if (!isItemStackEmpty(itemStack))
+        {
+            int slot;
+
+            if (itemStack.isItemDamaged())
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') NullPointerException might be thrown as 'itemStack' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2259)